### PR TITLE
Initial NodeJS Typescript binding and demo

### PR DESCRIPTION
**Pull request by ErisMik**

### CODE REVIEW

File: sdk/nodejs/src/picovoice.ts [hunk 1]
```diff
@@ -10,18 +10,36 @@
...
+  private readonly _version: string;
+  private readonly _porcupineVersion: string;
+  private readonly _rhinoVersion: string;
+  private readonly _contextInfo: string;
+
+  private isWakeWordDetected: boolean;
+
   /**
    * Creates an instance of Picovoice with a specific keyword and context.
    * @param {accessKey} AccessKey Obtained from the Picovoice Console (https://console.picovoice.ai/)
```
Comment (ksyeo1010) [inline sdk/nodejs/src/picovoice.ts:45]:
```suggestion
   * @param {string} accessKey Obtained from the Picovoice Console (https://console.picovoice.ai/)
```

File: sdk/nodejs/src/picovoice.ts [hunk 2]
```diff
@@ -10,18 +10,36 @@
...
  * Switches input from Porcupine to Rhino upon wake word detection, then back to Rhino upon inference conclusion.
  * Fires callbacks on wake word and inference events.
  */
-class Picovoice {
+export default class Picovoice {
+  private porcupine: Porcupine | null;
+  private rhino: Rhino | null;
+
+  private wakeWordCallback: WakeWordCallback;
+  private inferenceCallback: InferenceCallback;
```
Comment (ksyeo1010) [inline sdk/nodejs/src/picovoice.ts:32]:
```suggestion
  private readonly wakeWordCallback: WakeWordCallback;
  private readonly inferenceCallback: InferenceCallback;
```

File: sdk/nodejs/src/platforms.ts [hunk 3]
```diff
@@ -66,85 +66,77 @@ SYSTEM_TO_LIBRARY_PATH.set(
   `${PLATFORM_RASPBERRY_PI}/${ARM_CPU_CORTEX_A53}/pv_rhino.node`
 );
 SYSTEM_TO_LIBRARY_PATH.set(
-  `${SYSTEM_LINUX}/${ARM_CPU_CORTEX_A53}${ARM_CPU_64}`,
-  `${PLATFORM_RASPBERRY_PI}/${ARM_CPU_CORTEX_A53}${ARM_CPU_64}/pv_rhino.node`
+    `${SYSTEM_LINUX}/${ARM_CPU_CORTEX_A53}${ARM_CPU_64}`,
```
Comment (ksyeo1010) [inline sdk/nodejs/src/platforms.ts:69]: 2 spaces, same for rest of file. I think your editor adds  spaces automatically.

File: sdk/nodejs/test/index.test.ts [hunk 4]
```diff
@@ -10,15 +10,19 @@
...
 
-const { PvArgumentError, PvStateError } = require("./errors");
-const { getPlatform, getSystemLibraryPath } = require("./platforms");
+import * as fs from "fs";
+import { getInt16Frames, checkWaveFile } from "../src/wave_util";
+import { WaveFile } from "wavefile";
 
-const PICOVOICE_PORCUPINE_KEYWORD = 11;
+import { PicovoiceInvalidArgumentError, PicovoiceInvalidStateError } from "../src/errors";
+import { getPlatform, getSystemLibraryPath } from "../src/platforms";
```
Comment (ksyeo1010) [inline sdk/nodejs/test/index.test.ts:22]:
```suggestion
import { getPlatform } from "../src/platforms";
```

**Review by ksyeo1010 — CHANGES_REQUESTED** [review]
are the submodule updates also expected?

**Comment by ErisMik** [comment]
> are the submodule updates also expected?

Yes

File: sdk/nodejs/src/platforms.ts (hunk 3 above)
Comment (ErisMik) [reply sdk/nodejs/src/platforms.ts:69]: Ah, probably those were tab characters and my editor is set to convert tabs to 4 spaces

**Review by ksyeo1010 — APPROVED** [review]

**Review by mrrostam — APPROVED** [review]

**Pull request merged**